### PR TITLE
Validate extension failure behavior

### LINKED ISSUE #840: Agent should attempt fewer installs / updates of extensions
The agent can repeatedly try to install / disable / enable an extension even if doing so fails. It should try once and stop.

### PR DESCRIPTION
**Pull request by hglkrijger**
- since #840 has been opened, we have made some changes to how extension failures are handled
- add unit tests to validate these scenarios
- closes #840 

---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: tests/ga/test_extension.py [hunk 1]
```diff
@@ -769,6 +769,108 @@ def test_extensions_deleted(self, *args):
         self._assert_handler_status(protocol.report_vm_status, "Ready", 1, "1.0.1")
         self._assert_ext_status(protocol.report_ext_status, "success", 0)
 
+    @patch('subprocess.Popen.poll')
+    def test_install_failure(self, patch_poll, *args):
```
Comment (boumenot) [inline tests/ga/test_extension.py]: A method comment would make it clearer what you are actually trying to assert in this case.  The comment below makes that clearer, but a little verbiage would help.

File: tests/ga/test_extension.py [hunk 2]
```diff
@@ -769,6 +769,108 @@ def test_extensions_deleted(self, *args):
...
+        self.assertEqual(1, protocol.report_vm_status.call_count)
+        self._assert_handler_status(protocol.report_vm_status, "NotReady", expected_ext_count=0, version="1.0.0")
+
+        # Ensure subsequent no further retries are made
+        exthandlers_handler.run()
+        self.assertEqual(2, patch_poll.call_count)
+        self.assertEqual(2, protocol.report_vm_status.call_count)
+
+    @patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_enable_command')
+    def test_enable_failure(self, patch_get_enable_command, *args):
```
Comment (boumenot) [inline tests/ga/test_extension.py]: Ditto, method comment.

File: tests/ga/test_extension.py [hunk 3]
```diff
@@ -769,6 +769,108 @@ def test_extensions_deleted(self, *args):
...
+        self.assertEqual(1, patch_get_enable_command.call_count)
+        self.assertEqual(1, protocol.report_vm_status.call_count)
+        self._assert_handler_status(protocol.report_vm_status, "NotReady", expected_ext_count=1, version="1.0.0")
+
+        exthandlers_handler.run()
+        self.assertEqual(1, patch_get_enable_command.call_count)
+        self.assertEqual(2, protocol.report_vm_status.call_count)
+
+    @patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_disable_command')
+    def test_disable_failure(self, patch_get_disable_command, *args):
```
Comment (boumenot) [inline tests/ga/test_extension.py]: Ditto, method comment.

File: tests/ga/test_extension.py [hunk 4]
```diff
@@ -769,6 +769,108 @@ def test_extensions_deleted(self, *args):
...
+        self._assert_handler_status(protocol.report_vm_status, "NotReady", expected_ext_count=1, version="1.0.0")
+
+        # Ensure there are no further retries
+        exthandlers_handler.run()
+        self.assertEqual(1, patch_get_disable_command.call_count)
+        self.assertEqual(3, protocol.report_vm_status.call_count)
+        self._assert_handler_status(protocol.report_vm_status, "NotReady", expected_ext_count=1, version="1.0.0")
+
+    @patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_uninstall_command')
+    def test_uninstall_failure(self, patch_get_uninstall_command, *args):
```
Comment (boumenot) [inline tests/ga/test_extension.py]: Copy pasta.

File: tests/ga/test_extension.py (hunk 1 above)
Comment (hglkrijger) [reply tests/ga/test_extension.py]: Added some comments

File: tests/ga/test_extension.py [hunk 5]
```diff
@@ -769,6 +769,120 @@ def test_extensions_deleted(self, *args):
...
+        """
+        When extension install fails, the operation should not be retried.
+        """
+        test_data = WireProtocolData(DATA_FILE_EXT_SINGLE)
+        exthandlers_handler, protocol = self._create_mock(test_data, *args)
+
+        # Ensure initial install is unsuccessful
+        patch_poll.call_count = 0
+        patch_poll.return_value = 1
+        exthandlers_handler.run()
```
Comment (uiri) [inline tests/ga/test_extension.py]: I do not think this test is testing what you think it is testing.

`exthandlers_handler.run()` is called inside a `while self.running` loop inside UpdateHandler's `run` method. We should be testing that method to ensure that #840 is fixed.

**Pull request merged**